### ISSUE TITLE
Remove menu icons from non-document menu actions on macOS/OS X

### ISSUE THREAD

**Issue by follower** [LMMS/lmms#3349]
(_Originally this bug was this but see #3362:_ 

As previously mentioned (https://github.com/LMMS/lmms/issues/2890#issuecomment-273351599) the new white menu icons do not display correctly on Mac OS X 10.8 with Qt5:

![lmms-menu-icons-display-issue-qt5](https://cloud.githubusercontent.com/assets/189962/22863434/c9ec8d7e-f1a4-11e6-988d-14f14aa94810.png)

Related version info:

![lmms-menu-icons-display-issue-qt5-version-info](https://cloud.githubusercontent.com/assets/189962/22863458/36929612-f1a5-11e6-9e9a-a4803add1295.png)

I assume if they display on later OS X versions it's due to a change in menu background colour?
)

_Now:_

FWIW, I think on Mac OS X / macOS icons should not be displayed on menus at all, as it's not "standard practice" on the platform and IMO makes lmms appear "less native". Further details on request. :)

It appears specifying [`Qt::AA_DontShowIconsInMenus`](http://doc.qt.io/qt-5/qt.html#ApplicationAttribute-enum) would enable this but I've not tested it yet.

**Comment by tresf** [comment]
> IMO makes lmms appear "less native"

Probably the more important part of this is the UI experience.  OSs which can't align text in a menu should have the icons turned off, period.  Please issue a Pull Request with `Qt::AA_DontShowIconsInMenus` activated for MacOS only and we can finally put this issue to rest. 👍

**Comment by tresf** [comment]
Sounds good.  I noticed the icons don't seem so cosmetically weird if the icons aren't white.  For reference here's a side-by-side with Google Chrome:

![image](https://cloud.githubusercontent.com/assets/6345473/22896104/da46428a-f1ec-11e6-9ffb-39a09d87799d.png)  ![image](https://cloud.githubusercontent.com/assets/6345473/22896001/77de8684-f1ec-11e6-9196-fe4c9c8f5b70.png)

**Comment by tresf** [comment]
I just read a bug report with users complaining about this identical behavior, except when in dark mode.  Apparently there's an unmerged feature to help... https://codereview.qt-project.org/#/c/115120/

Linked from https://bugreports.qt.io/browse/QTBUG-42109.

For the short-term, we can use a quick hack, such as inverting the color palette for these icons. http://doc.qt.io/qt-5/qimage.html#invertPixels

**Comment by follower** [comment]
I managed to get the submenu icons to display (still as white icons for default theme, as icons for classic theme) by using `setIconVisibleInMenu(true)` but ([surprise](https://bugreports.qt.io/browse/QTBUG-4424), [surprise](https://bugreports.qt.io/browse/QTBUG-44565)) there appears to be a bug in Qt that means the change in visibility doesn't propagate correctly (in spite of the fact that the original issue was marked as "Cannot Reproduce") apparently when `addAction()` is used.

I've worked out a few variations of a workaround for the apparent bug and plan to use the one which is least disruptive to the existing code and requires least additional code.

**Comment by follower** [comment]
So, once `QApplication::setAttribute(Qt::AA_DontShowIconsInMenus, true);` has been used, in theory we can turn individual icons back on using `setIconVisibleInMenu(true) `.

Let's look at the different approaches we can use by using this example from `MainWindow::updateRecentlyOpenedProjectsMenu()` in `src/gui/MainWindow.cpp`:
```
m_recentlyOpenedProjectsMenu->addAction(embed::getIconPixmap( "project_file" ), *it );
```

### Making icons visible
#### Chained method calls
The simplest approach would be to use:

```
m_recentlyOpenedProjectsMenu->addAction(embed::getIconPixmap( "project_file" ), *it )->setIconVisibleInMenu(true);
```

However, I wasn't sure if "chained" method calls like that were within the lmms "style" (it's seems they're used sometimes but not frequently) and while (in theory) it should be a no-op for other platforms, since we're wanting to make this change be Apple only it would be a little "messy" to make the compilation conditional.

That consideration becomes a somewhat moot point however due to the bug mentioned in my previous comment.

#### Using `actions().last()`
With the bug in mind and the desire for conditional compilation I think the preferable approach is to add this as a separate line:
```
m_recentlyOpenedProjectsMenu->actions().last()->setIconVisibleInMenu(true);
```

#### Using the returned object
Another alternative would be to avoid the `actions()`/`last()` method calls and do this but again it's a more complex change:
```
QAction *newAction = m_recentlyOpenedProjectsMenu->addAction(embed::getIconPixmap( "project_file" ), *it );
newAction->setIconVisibleInMenu(true)
```

#### Creating the `QAction` first and avoiding the bug/workaround
AFAICT the bug occurs when using `.addAction()` rather than creating the `QAction` first, so we could change the existing code to be something like this but IMO doing this creates too many changes in the existing code:
```
QAction *newAction = new QAction(embed::getIconPixmap( "project_file" ), *it );
newAction->setIconVisibleInMenu(true);
m_recentlyOpenedProjectsMenu->addAction(newAction);
```

### The bug
The change in icon visibility doesn't get propagated (using `sendDataChanged()`) due to the test(s) here: <https://github.com/qt/qtbase/blob/99b6eb4382e135815b53b567297be9cd4ecfd96a/src/widgets/kernel/qaction.cpp#L1278>
```
        // Only send data changed if we really need to.
        if (oldValue != -1
            || (oldValue == -1
                && visible == !QApplication::instance()->testAttribute(Qt::AA_DontShowIconsInMenus))) {
            d->sendDataChanged();
        }
```
It's a little unclear what the correct test should be but that discussion belongs in the bug report ["[QTBUG-44565] CLONE - QAction::setIconVisibleInMenu does not work."](https://bugreports.qt.io/browse/QTBUG-44565)

### Bug workarounds
#### Toggle approach
While using the `actions().last()` approach works for our purpose, due to the bug this change in visibility doesn't get propagated so we need to workaround it by doing setting the value to `false` first (when first instantiated the value is `-1` i.e. default/global setting):
```
m_recentlyOpenedProjectsMenu->actions().last()->setIconVisibleInMenu(false);
m_recentlyOpenedProjectsMenu->actions().last()->setIconVisibleInMenu(true);
```

#### Manual event propagation trigger
Doing the above works around the bug in the "quickest" way but doesn't actually result in the same flow of events.

To actually cause the "correct" event flow to happen we could manually trigger the event ourselves like this but this seems overkill for our purposes:
```
m_recentlyOpenedProjectsMenu->actions().last()->setIconVisibleInMenu(true);
QActionEvent::QActionEvent event(QEvent::ActionChanged, m_recentlyOpenedProjectsMenu->actions().last());
QApplication::sendEvent(m_recentlyOpenedProjectsMenu, &event);
```

### Conclusion
Anyway, this is mostly a record for my benefit but might be useful in case other people encounter the same issue.

PR with functionality + fix will be forthcoming.

**Comment by follower** [comment]
Would've been a good idea to create a separate issue earlier but oh well... :)

#3362 is now for handling icon inversion but I'm not planning to touch that for the moment.

This is now just for removing the menu icons.

**Comment by follower** [comment]
Here's how the menus look with the patches...

#### Classic theme (projects & templates submenus):
![lmms-menu-icons-mac-fix-classic-templates-only](https://cloud.githubusercontent.com/assets/189962/23034049/48883650-f4df-11e6-84d7-cae55eb65e74.png)
![lmms-menu-icons-mac-fix-classic-projects-only](https://cloud.githubusercontent.com/assets/189962/23034047/487d18b0-f4df-11e6-8659-51cca679675b.png)

#### Default theme (projects & templates submenus):
![lmms-menu-icons-mac-fix-default-templates-only](https://cloud.githubusercontent.com/assets/189962/23034048/487fd7da-f4df-11e6-9219-ec4bdf7c78c7.png)
![lmms-menu-icons-mac-fix-default-projects-only](https://cloud.githubusercontent.com/assets/189962/23034050/48887688-f4df-11e6-97ad-b1870ba36a87.png)

**Comment by follower** [comment]
Closed by #3363.
